Add klap support for TAPO protocol by splitting out Transports and Protocols

sdb9696 opened the pull request:
This PR adds KLAP support for the TAPO/SMART protocol.  The way the encryption works is that the the original message that worked with https://github.com/python-kasa/python-kasa/pull/552 is instead encrypted with KLAP:

```
"method": method,
"params": params,
"requestID": self.request_id_generator.generate_id(),
"request_time_milis": round(time.time() * 1000),
"terminal_uuid": self.terminal_uuid,
```
As the current klapprotocol implementation assumes the old kasa style of request/response this PR proposes to separate the transport from the protocol.  The changes are:
1. Introduce a base TPLinkTransport abstract class.
2. Rename `TplinkKlap` to `TPlinkIotProtocol`.  It seems this is the distinction in the discovery result, device_type beginning with `IOT.` uses the old protocol, device_type beginning with `SMART.` uses the new.
3. Rename `TPLinkAes` to `TPLinkSmartProtocol`.
4. Break out klap logic from `TPLinkKlap` into a new `TPLinkKlapTransport` class.
5. Break out AES logic from `TPLinkAes` into a new `TPLinkAesTransport` class.
6. The klap handshake hashes are slightly different with the `SMART.` devices so there's a TPLinkKlapTransportV2 class derived from TPLinkKlapTransport that overrides the three hash functions.

A good starting point to understand the changes here should be the `get_protocol_from_connection_name` [diff in device_factory.py](https://github.com/python-kasa/python-kasa/pull/557/files#diff-74ee482e62e8aa8c6f3d73030170d5c3422523dbb0f97254a201ce897da61c5a)

rytilahti reviewed (commented):
Just a couple of very brief comments, mostly about things to make it easier to review.

sdb9696 commented on kasa/aestransport.py:
```diff
@@ -98,14 +84,14 @@ def hash_credentials(self, credentials, try_login_version2):
         if try_login_version2:
             un = base64.b64encode(
-                _sha1(credentials.username.encode()).encode()
+                _sha1_hex(credentials.username.encode()).encode()
```
Removed

sdb9696 commented on devtools/dump_devinfo.py:
```diff
@@ -7,6 +7,7 @@
 Executing this script will several modules and methods one by one,
```
Sure, I'll create a separate PR for this.

sdb9696 replied:
Ok that's done https://github.com/python-kasa/python-kasa/pull/561

sdb9696 commented:
Tests have now been added and the test framework is working with the new tapo devices.  I will add more tests but wanted to get this in to ease review.

rytilahti commented on kasa/protocol.py:
```diff
@@ -32,6 +33,54 @@
+    return hash
+
+
+class TPLinkTransport(ABC):
```
```suggestion
class Transport(ABC):
```
or `TransportProtocol`? I'd rather like to see us getting rid off superfluous prefixes if possible, or at least try to not add more of them.

sdb9696 replied:
Done, renamed BaseTransport

rytilahti commented on kasa/aestransport.py:
```diff
@@ -243,96 +241,23 @@ def handshake_session_expired(self):
-        return response_data
+    async def send(self, request: str):
+        """Send the request."""
+        if self.needs_handshake():
+            raise SmartDeviceException(
+                "Handshake must be complete before trying to send"
+            )
+        if self.needs_login():
+            raise SmartDeviceException("Login must be complete before trying to send")
```
Just a question, would it make sense to perform handshaking (and login, too?) transparently whenever needed?

rytilahti commented on kasa/iotprotocol.py:
```diff
@@ -0,0 +1,106 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class TPLinkIotProtocol(TPLinkProtocol):
```
```suggestion
class IotProtocol(TPLinkProtocol):
```
We probably want to have some documentation (or at least docstrings) explaining what different protocols mean in the context of the lib, considering IoT is rather generic term :-)

sdb9696 replied:
Done

rytilahti commented on kasa/iotprotocol.py:
```diff
@@ -0,0 +1,106 @@
+    ) -> None:
+        super().__init__(host=host, port=self.DEFAULT_PORT)
+
+        self.credentials: Credentials = (
+            credentials
+            if credentials and credentials.username and credentials.password
+            else Credentials(username="", password="")
+        )
+        self.transport: TPLinkTransport = (
+            transport
+            if transport
+            else TPLinkKlapTransport(
+                host, credentials=self.credentials, timeout=timeout
+            )
+        )
```
See my other comment wrt. using a simple `or` for these.

sdb9696 replied:
Done

rytilahti commented on kasa/device_factory.py:
```diff
@@ -129,3 +133,37 @@ def get_device_class_from_info(info: Dict[str, Any]) -> Type[SmartDevice]:
+    """Return the protocol from the connection name."""
+    supported_device_protocols: dict[
+        str, Tuple[Type[TPLinkProtocol], Type[TPLinkTransport]]
+    ] = {
+        "IOT.KLAP": (TPLinkIotProtocol, TPLinkKlapTransport),
+        "SMART.AES": (TPLinkSmartProtocol, TPLinkAesTransport),
+        "SMART.KLAP": (TPLinkSmartProtocol, TPlinkKlapTransportV2),
```
Tbf, I don't like the usage of a "magic" string to proto & transport mapping, but that's something we may need to figure out in another PR to keep this tidier?

sdb9696 replied:
Ok agreed.  Let's add the `ConnectionParameters` in the next PR

rytilahti commented on kasa/device_factory.py:
```diff
@@ -129,3 +133,37 @@ def get_device_class_from_info(info: Dict[str, Any]) -> Type[SmartDevice]:
+    if connection_name not in supported_device_protocols:
+        return None
+
+    protocol_transport_tuple = supported_device_protocols.get(connection_name)
+    transport: TPLinkTransport = protocol_transport_tuple[1](  # type: ignore
+        host, credentials=credentials
+    )
+    protocol: TPLinkProtocol = protocol_transport_tuple[0](  # type: ignore
+        host, credentials=credentials, transport=transport
+    )
```
```suggestion
    protocol_class, transport_class = supported_device_protocols.get(connection_name)
    transport: TPLinkTransport = transport_class(  # type: ignore
        host, credentials=credentials
    )
    protocol: TPLinkProtocol = protocol_class(  # type: ignore
        host, credentials=credentials, transport=transport
    )
```
Also, are the type ignores necessary?

sdb9696 replied:
They were necessary because mypy didn't recognise the null check beforehand.  Now the type:ignore is moved to line  line 162 which is done as per comment

rytilahti reviewed (changes requested):
That's quite a bit of code to review and test, but here's the first round of comments. In general, separating the concerns of the encryption and the transport sounds good to me :+1:

A couple of things:
* Update the PR title to be more descriptive.
* Class naming: let's try to avoid useless `TPLink*`, `Smart` etc. prefixes where feasible, that pattern came from the pyhs100 project to keep backwards compatibility, but we should not introduce any more of them.
* Using a self-created str<->(protocol, transport) mapping feels wrong and potentially error-prone when new versions will be deployed. I think we should decide soon what's a good abstraction for the discovery-like information to device instances..

sdb9696 commented on kasa/tests/test_smartdevice.py:
```diff
@@ -22,11 +22,13 @@
 ]
 
 
+@device_iot
```
Thanks for the review @rytilahti. I know it was a lot and I thought hard about how to make it smaller but couldn’t come up with anything that made much sense. 

Most of the comments are agreeable and doable and I will hopefully get sorted Monday. There is one question I have to clarify with you first. I restricted these tests with `@device_iot` as a short term measure to keep the PR from getting bigger (yes I know :-). Currently too many tests access sysinfo and break. 

So I think we have two options:

1. Move all the SMART fixtures into a different directory so they don’t run on all tests.
2. Restrict them for now with `@device_iot` and then gradually re-enable test_file by test_file as we introduce the interfaces. 

I prefer 2. because it will mean we have a single test suite and a good interface. What do you think?

rytilahti replied:
Hmm, so those tests that test for specific payloads (like those schema checks in test_smartdevice) should only be tested against "device_iot"/"legacy" payloads, and indeed worth guarding for the time being.

Even then, we probably want to keep the "smart" fixtures separated from the legacy ones, so creating a new directory to host them from the get-go is a good idea if that's simple to achieve. If not, let's keep it simple for the time being and revisit it later.

But maybe the more generic ones (like testing for state_attributes structure, tests for cli outputs, ..) could still test everything if doing so doesn't require too many changes?

sdb9696 commented on kasa/tests/test_cli.py:
```diff
@@ -6,19 +6,23 @@
+from .newfakes import FakeSmartProtocol, FakeTransportProtocol
 
 
+@device_iot
```
Agreed and a few of them I am doing this but I need to iterate the test framework a bit more to get this working for all cli commands for the new protocol

sdb9696 commented on kasa/tests/newfakes.py:
```diff
@@ -285,6 +287,39 @@ def success(res):
 }
 
 
+class FakeSmartProtocol(TPLinkSmartProtocol):
```
Ok

sdb9696 commented on kasa/tests/conftest.py:
```diff
@@ -55,37 +60,46 @@
-            if filt in basename(dev):
+            model_name = filt
+            if is_smart_protocol:
+                model_name = model_name + ".smart"
```
Done

sdb9696 commented:
HI @rytilahti, I had a bit of time today to address this review.  There's still work needed on the test framework but I think it will be better to perfect that in further PRs.  After we get this in we can hopefully start to focus on separate streams, i.e. TAPO PRs, test framework PRs, test coverage PRs, fixture PRs and interface PRs.  I hope we have the baseline now :)

rytilahti commented on kasa/protocol.py:
```diff
@@ -41,6 +90,7 @@ def __init__(
         *,
         port: Optional[int] = None,
         credentials: Optional[Credentials] = None,
+        transport: Optional[BaseTransport] = None,
```
I realized that not all implementations accept `transport` (and at least `TPLinkSmartHomeProtocol` has `timeout` added).

Maybe we should enforce consistent parameters all across the board (by adding a test for the ctors?)?

sdb9696 replied:
I agree it makes sense.  Would prefer to add it in a subsequent PR as it'll be easier to see what's going and we may want to convert the legacy TPLinkSmartHomeProtocol class to use the IotProtocol with it's own XoR transport broken out.  Am thinking about the request for re-use etc.

rytilahti replied:
Agreed, let's get this one merged and follow-up with smaller and easier to review PRs :-)

sdb9696 commented on kasa/protocol.py:
```diff
@@ -32,6 +33,54 @@
+
+
+class TPLinkTransport(ABC):
+    """Base class for all TP-Link KASA-KLAP and TAPO transports."""
```
Done

sdb9696 commented on kasa/smartprotocol.py:
```diff
@@ -0,0 +1,223 @@
+
+        self.credentials: Credentials = (
+            credentials
+            if credentials and credentials.username and credentials.password
```
Done

sdb9696 commented on kasa/smartprotocol.py:
```diff
@@ -0,0 +1,223 @@
+        self.transport: TPLinkTransport = (
+            transport
+            if transport
+            else TPLinkAesTransport(host, credentials=self.credentials, timeout=timeout)
```
Done

sdb9696 commented on kasa/aestransport.py:
```diff
@@ -132,55 +118,70 @@ async def client_post(self, url, params=None, data=None, json=None, headers=None
+            await self.perform_login(request, False)
+        except AuthenticationException:
+            await self.perform_handshake()
+            await self.perform_login(request, True)
```
Done

sdb9696 commented on kasa/aestransport.py:
```diff
@@ -132,55 +118,70 @@ async def client_post(self, url, params=None, data=None, json=None, headers=None
+
+    def needs_handshake(self) -> bool:
+        """Return true if the transport needs to do a handshake."""
+        return not self.handshake_done or self.handshake_session_expired()
```
Done

sdb9696 commented on kasa/aestransport.py:
```diff
@@ -132,55 +118,70 @@ async def client_post(self, url, params=None, data=None, json=None, headers=None
-        return request
-
-    async def perform_login(self, login_v2):
+    async def perform_login(self, login_request, login_v2):
```
Done

sdb9696 commented on kasa/tests/test_device_factory.py:
```diff
@@ -70,32 +85,52 @@ async def test_connect_logs_connect_time(
+        mocker.patch("kasa.SmartProtocol.query", return_value=all_fixture_data)
+
+        dr = DiscoveryResult(**discovery_info["result"])
+        connection_name = (
+            dr.device_type.split(".")[0] + "." + dr.mgt_encrypt_schm.encrypt_type
+        )
```
Will resolve this comment as we're agreed on ConnectionParams in subsequent PR

sdb9696 commented on kasa/aestransport.py:
```diff
@@ -132,55 +118,70 @@ async def client_post(self, url, params=None, data=None, json=None, headers=None
         self.login_token = None
 
+        if isinstance(login_request, str):
+            login_request = json_loads(login_request)
```
I seem to recall I was originally trying to use dicts but it got complicated because `transport.send()` has to take a string because the `SmartProtocol.query()` takes a method name rather than a dict.  We could probably improve this down the line so we have a more concrete implementation of a `query_request`.

sdb9696 commented on kasa/protocol.py:
```diff
@@ -32,6 +33,54 @@
+
+    @abstractmethod
+    def needs_login(self) -> bool:
+        """Return true if the transport needs to do a login."""
```
Done

rytilahti commented on kasa/tests/test_plug.py:
```diff
@@ -28,3 +28,14 @@ async def test_led(dev):
     await dev.set_led(original)
+
+
+@plug_smart
```
Yup, we can do that, too. On top of that we _may_ want to have tests for the response payloads, but that's something for another day.

sdb9696 commented:
> This is looking good, I added some extra comments on top of the few already existing unresolveds. @sdb9696 take a look and see what makes sense to fix prior merging this so that we can start again having smaller PRs :-)

Hi @rytilahti , I've addressed most of the comments from today and the open ones I missed from the other day (for some reason they keep becoming hidden for me in github).

If we're good to merge this now I'll go through afterwards and collate all of the "Subsequent PR" items we discussed and add them to the [discussion](https://github.com/python-kasa/python-kasa/discussions/559) so we don't lose sight of them.
